File: engineering/network/03-io-multiplexing/skeleton/python/main.py
# 模型：select.select（单线程多路复用）
import socket
import select
import threading
import time

def run_server():
    # TODO: 创建 TCP socket，绑定 0 端口，listen
    # TODO: 使用 select.select 监控 listener 和所有 client sockets
    # TODO: 新连接 -> accept 并加入监控列表
    # TODO: 可读连接 -> recv 数据，echo 回去；如果 recv 返回空 bytes，关闭并移除连接
    # TODO: 在后台线程中运行上述循环
    # TODO: 返回实际地址 (host, port)
    pass

def run_client(addr, msg):
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.settimeout(1)
    s.connect(addr)
    s.sendall(msg.encode())
    data = s.recv(1024)
    s.close()
    return data.decode()

def main():
    addr = run_server()
    time.sleep(0.1)

    results = []
    for i in range(3):
        msg = f"client{i}"
        reply = run_client(addr, msg)
        results.append((msg, reply))

    for msg, reply in results:
        if reply != msg:
            print(f"FAIL: expected {msg!r} got {reply!r}")
            return
    print("PASS: multiplexed echo server correct")

if __name__ == "__main__":
    main()

File: engineering/network/03-io-multiplexing/skeleton/go/main.go
package main

import (
	"fmt"
	"net"
	"time"
)

// runServer 启动单 goroutine TCP echo server（使用 net + channel 模拟多路复用）。
func runServer() (string, error) {
	// TODO: 监听 TCP 0 端口
	// TODO: 在单个 goroutine 中：
	//       1. 使用 listener.Accept() 获取新连接（设置 deadline 避免永久阻塞）
	//       2. 维护一个连接列表
	//       3. 对列表中的每个连接，尝试读取数据（设置 read deadline）
	//       4. 如果有数据，echo 回去；如果连接关闭，从列表中移除
	//       5. 循环直到收到退出信号
	return "", fmt.Errorf("not implemented")
}

// runClient 连接 server，发送 msg，等待回复。
func runClient(addr string, msg string) (string, error) {
	conn, err := net.Dial("tcp", addr)
	if err != nil {
		return "", err
	}
	defer conn.Close()
	if _, err := conn.Write([]byte(msg)); err != nil {
		return "", err
	}
	buf := make([]byte, 1024)
	conn.SetReadDeadline(time.Now().Add(500 * time.Millisecond))
	n, err := conn.Read(buf)
	if err != nil {
		return "", err
	}
	return string(buf[:n]), nil
}

func main() {
	addr, err := runServer()
	if err != nil {
		fmt.Printf("FAIL: server start: %v\n", err)
		return
	}
	time.Sleep(100 * time.Millisecond)

	// 并发启动 3 个 client
	results := make(chan string, 3)
	for i := 0; i < 3; i++ {
		go func(id int) {
			msg := fmt.Sprintf("client%d", id)
			reply, err := runClient(addr, msg)
			if err != nil {
				results <- fmt.Sprintf("FAIL client%d: %v", id, err)
				return
			}
			if reply != msg {
				results <- fmt.Sprintf("FAIL client%d: expected %q got %q", id, msg, reply)
				return
			}
			results <- "OK"
		}(i)
	}

	for i := 0; i < 3; i++ {
		r := <-results
		if r != "OK" {
			fmt.Println(r)
			return
		}
	}
	fmt.Println("PASS: multiplexed echo server correct")
}
